tuf/api: Expose tuf.api as a package (take 2)

joshuagl opened the pull request:
Taking over from #1157 

**Description of the changes being introduced by the pull request**:
* add tuf/api/__init__.py, turning tuf/api into a full package for distribution purposes
* add a pylintrc for tuf/api and update tox to use it (and ignore tuf/api with the default pylintrc)
* minor stylistic changes to tuf/api/metadata.py to resolve some pylint warnings

Note: ac8d273 disables a single pylintrc check in a code comment. I was loath to do this, but I think it's better to explicitly disable that site than disable the message for all code before we've decided on coding style.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

woodruffw committed:
tuf/api: Expose tuf.api as a package

Signed-off-by: William Woodruff <william@trailofbits.com>

joshuagl committed:
Add minimal pylintrc for new code in tuf/api

Add a minimal pylintrc to lint for new code being developed in tuf/api and
update the tox configuration to ignore tuf/api with the default pylintrc
and run an extra invocation of pylint for just the modules in tuf/api.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Remove unused import from tuf.api.metadata

The logging module is not used in metadata, therefore remove it

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Remove use of single letter variable

A single letter variable name of 'f' causes pylint to throw a coding style
convention warning:

C0103: Variable name "f" doesn't conform to snake_case naming style
(invalid-name)

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Remove else after raise usage in api.metadata

Using an else after a raise results in a refactor message from pylint:

R1720: Unnecessary "elif" after "raise" (no-else-raise)

This is because the raise will exit the block, and pylint suggests that
explicit if's, rather than an if-elif-else, are clearer style. Update the
style of Metadata.verify() to match pylint expectations.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Disable an instance of too-many-arguments error

The Targets constructor takes seven arguments, which violates pylints
default value of five for max-arguments:

R0913: Too many arguments (7/5) (too-many-arguments)

As this feels like a coding style decision that should be made and
documented disable that test for only the Targets constructor until
a coding style decision has been made and documented as a decision
record.

Signed-off-by: Joshua Lock <jlock@vmware.com>

lukpueh commented on tox.ini:
```diff
@@ -14,7 +14,8 @@ skipsdist = true
 commands =
-    pylint {toxinidir}/tuf
+    pylint {toxinidir}/tuf --ignore={toxinidir}/tuf/api
+    pylint {toxinidir}/tuf/api --rcfile={toxinidir}/tuf/api/pylintrc
```
What about https://github.com/theupdateframework/tuf/pull/1157#issuecomment-702025224? Or do you purposefully want to be explicit here?

lukpueh replied:
Also [looks](https://travis-ci.org/github/theupdateframework/tuf/builds/736003163) like we can only run pylint on supported versions for `tuf.api`.

lukpueh replied:
I wonder if the linter really needs to run for every supported version. In [in-toto we have a separate lint tox env](https://github.com/in-toto/in-toto/blob/c775e3b260a50ec5fb673a9b725e2d2eacedfb6b/tox.ini#L22-L25) that uses the latest supported Python version.

joshuagl replied:
> What about [#1157 (comment)](https://github.com/theupdateframework/tuf/pull/1157#issuecomment-702025224)? Or do you purposefully want to be explicit here?

It didn't seem to work when I tried it. May be because cwd isn't a Python module when we invoke pylint? Being explicit seemed easier than relying on pylint magic I couldn't fully decipher from the documentation.

> I wonder if the linter really needs to run for every supported version. In in-toto we have a separate lint tox env that uses the latest supported Python version.

This feels like a sensible option, I'll add a separate lint env.

lukpueh commented on tuf/api/pylintrc:
```diff
@@ -0,0 +1,9 @@
+[BASIC]
+good-names=e, fn, fp
```
Minor nit: Right now we don't use those names. What do you think about removing the line and only add when needed, or when fleshing out the config in order to align with the style guide TBD (https://github.com/secure-systems-lab/code-style-guidelines/issues/20)?

joshuagl replied:
Completely agree, I'll remove this line.

lukpueh reviewed (changes requested):
Thanks a lot, @joshuagl! Looks mostly good to me.

joshuagl committed:
Move linting to a separate tox env

We don't need to lint the code with every version of Python, instead add
an extra tox env which lints once with the latest supported Python version

Signed-off-by: Joshua Lock <jlock@vmware.com>

lukpueh commented on tox.ini:
```diff
@@ -39,3 +38,8 @@ deps =
     coverage run aggregate_tests.py
     coverage report -m
+
+[testenv:lint]
```
We also need to [adopt this change in .travis.yml](https://github.com/in-toto/in-toto/blob/c775e3b260a50ec5fb673a9b725e2d2eacedfb6b/.travis.yml#L20-L21). And should we maybe move bandit to the lint env too?

joshuagl replied:
Ugh, I always forget to touch `.travis.yml` – thanks!

joshuagl committed:
Switch sslib master env to Python 3.8

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
I've force pushed an update to squash fixups, which prevents the DCO from complaining. 

You can see the pre rebase history here https://github.com/theupdateframework/tuf/compare/develop...joshuagl:ww/tuf-api-package-hist where I've added 81e0ef6 to add the lint env to travis, and move bandit into the lint env. I also changed 427759b while I was in the .travis.yml to use newer Python for with-sslib-master

woodruffw commented:
Thanks @joshuagl!

Commit:
ci: skip fossa and coverage for lint jobs

Authored-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: Joshua Lock <jlock@vmware.com>

lukpueh reviewed (approved):
🎉
